lib/base-compiler.js

```diff
@@ -185,22 +208,41 @@ Compile.prototype.compile = function (source, options, filters) {
             options = self.prepareArguments(options, filters, inputFilename, outputFilename);
 
             options = options.filter(_.identity);
-            return self.runCompiler(self.compiler.exe, options, self.filename(inputFilename))
-                .then(function (result) {
-                    result.dirPath = dirPath;
-                    if (result.code !== 0) {
-                        result.asm = "<Compilation failed>";
-                        return result;
+
+            var asmPromise = self.runCompiler(self.compiler.exe, options, self.filename(inputFilename));
+
+            var astPromise;
+            if (backendOptions.produceAst) {
+                astPromise = self.generateAST(inputFilename, options);
+            }
+            else {
+                astPromise = Promise.resolve("");
+            }
+            
+            return Promise.all([asmPromise, astPromise])
+                .then(function (results) {
+                    var asmResult = results[0];
+                    var astResult = results[1];
+
+                    asmResult.dirPath = dirPath;
+                    if (asmResult.code !== 0) {
+                        asmResult.asm = "<Compilation failed>";
+                        return asmResult;
                     }
-                    result.hasOptOutput = false;
+                    asmResult.hasOptOutput = false;
                     if (self.couldHaveOptOutput(options, self.compiler.version)) {
                         const optPath = path.join(dirPath, outputFilebase + ".opt.yaml");
                         if (fs.existsSync(optPath)) {
-                            result.hasOptOutput = true;
-                            result.optPath = optPath;
+                            asmResult.hasOptOutput = true;
+                            asmResult.optPath = optPath;
                         }
                     }
-                    return self.postProcess(result, outputFilename, filters);
+                    if (astResult) {
+                        asmResult.hasAstOutput = true;
```

Comment:
just FYI i am removing hasOptOutput in a coming patch. I realised i can just check for astOutput != ""